File: src/keboola_mcp_server/tools/sql.py
import csv
import logging
from io import StringIO
from typing import Annotated

from fastmcp import Context, FastMCP
from fastmcp.tools import FunctionTool
from mcp.types import ProgressNotification, ProgressNotificationParams, ProgressToken, ToolAnnotations
from pydantic import BaseModel, Field

from keboola_mcp_server.errors import tool_errors
from keboola_mcp_server.workspace import JobSubmittedInfo, SqlSelectData, WorkspaceManager

LOG = logging.getLogger(__name__)

SQL_TOOLS_TAG = 'sql'
MAX_ROWS = 1_000
MAX_CHARS = 50_000


def _client_progress_token(ctx: Context) -> ProgressToken | None:
    """Returns the progress token the client included in the original `tools/call`, or None.

    Per MCP spec, a server may only send `notifications/progress` for a request when the client
    explicitly provided a `progressToken` in that request's `_meta`. Tools that fall through here
    without a token must stay silent — emitting unsolicited progress can confuse strict clients.
    """
    rc = ctx.request_context
    if rc is None or rc.meta is None:
        return None
    return rc.meta.progressToken


async def _emit_job_submitted_progress(ctx: Context, progress_token: ProgressToken, info: JobSubmittedInfo) -> None:
    """Surfaces the backend job handle to the client so it can cancel out-of-band by POSTing to
    `info.cancellation_url`. The structured data lives under `params._meta`; the human-readable
    `message` is for clients that surface progress as text only and ignore `_meta`.
    """
    params = ProgressNotificationParams.model_validate(
        {
            'progressToken': progress_token,
            'progress': 0,
            'message': f'Submitted to {info.backend}',
            '_meta': {
                'keboola.queryJobId': info.job_id,
                'keboola.backend': info.backend,
                # `cancellation_url` may be None when a backend does not expose an out-of-band
                # cancel endpoint; clients should treat the field as optional.
                'keboola.cancellationUrl': info.cancellation_url,
            },
        }
    )
    await ctx.send_notification(ProgressNotification(method='notifications/progress', params=params))


class QueryDataOutput(BaseModel):
    """Output model for SQL query results."""

    query_name: str = Field(description='The name of the executed query')
    csv_data: str = Field(description='The retrieved data in CSV format')
    message: str | None = Field(default=None, description='A message from the query execution')


def add_sql_tools(mcp: FastMCP) -> None:
    """Add tools to the MCP server."""
    mcp.add_tool(
        FunctionTool.from_function(
            query_data,
            annotations=ToolAnnotations(readOnlyHint=True),
            tags={SQL_TOOLS_TAG},
        )
    )
    LOG.info('SQL tools added to the MCP server.')


@tool_errors()
async def query_data(
    sql_query: Annotated[str, Field(description='SQL SELECT query to run.')],
    query_name: Annotated[
        str,
        Field(
            description=(
                'A concise, human-readable name for this query based on its purpose and what data it retrieves. '
                'Use normal words with spaces (e.g., "Customer Orders Last Month", "Top Selling Products", '
                '"User Activity Summary").'
            )
        ),
    ],
    ctx: Context,
) -> QueryDataOutput:
    """
    Executes an SQL SELECT query to get the data from the underlying database.

    BEFORE QUERYING:
    * Always verify the table has a non-null fullyQualifiedName from get_tables tool.
      If it does not, the table is not SQL-accessible from this workspace — do not attempt the query and inform user.

    CRITICAL SQL REQUIREMENTS:

    * ALWAYS check the SQL dialect before constructing queries.
    * Do not include any comments in the SQL code
    * Use delimited identifiers and FQN format for the current SQL dialect.

    TABLE AND COLUMN REFERENCES:
    * Always use fully qualified table names in the exact FQN format provided by table information tools
    * Follow the identifier structure exactly as shown by table info tools for the current SQL dialect
    * Always use delimited identifiers when referring to table columns

    CTE (WITH CLAUSE) RULES:
    * ALL column references in main query MUST match exact case used in the CTE
    * If you alias a column in a CTE, reference it under the aliased name in the subsequent queries
    * Define all column aliases explicitly in CTEs
    * Use delimited identifiers in both CTE definition and references to preserve case

    FUNCTION COMPATIBILITY:
    * Check data types before using date functions (DATE_TRUNC, EXTRACT require proper date/timestamp types)
    * Cast VARCHAR columns to appropriate types before using in date/numeric functions

    ERROR PREVENTION:
    * Never pass empty strings ('') where numeric or date values are expected
    * Use NULLIF or CASE statements to handle empty values
    * Always use TRY_CAST or similar safe casting functions when converting data types
    * Check for division by zero using NULLIF(denominator, 0)
    * Always use the LIMIT clause in your SELECT statements when fetching data. There are hard limits imposed
      by this tool on the maximum number of rows that can be fetched and the maximum number of characters.
      The tool will truncate the data if those limits are exceeded.

    DATA VALIDATION:
    * When querying columns with categorical values, use query_data tool to inspect distinct values beforehand
    * Ensure valid filtering by checking actual data values first
    """
    workspace_manager = WorkspaceManager.from_state(ctx.session.state)

    progress_token = _client_progress_token(ctx)

    async def _on_job_submitted(info: JobSubmittedInfo) -> None:
        await _emit_job_submitted_progress(ctx, progress_token, info)

    result = await workspace_manager.execute_query(
        sql_query,
        max_rows=MAX_ROWS,
        max_chars=MAX_CHARS,
        on_job_submitted=_on_job_submitted if progress_token is not None else None,
    )
    LOG.info(' '.join(filter(None, [f'Query "{query_name}" executed successfully.', result.message])))
    if result.is_ok:
        if result.data:
            data = result.data
        else:
            # non-SELECT query, this should not really happen, because this tool is for running SELECT queries
            data = SqlSelectData(columns=['message'], rows=[{'message': result.message}])

        # Convert to CSV
        output = StringIO()
        writer = csv.DictWriter(output, fieldnames=data.columns)
        writer.writeheader()
        writer.writerows(data.rows)

        return QueryDataOutput(query_name=query_name, csv_data=output.getvalue(), message=result.message)

    else:
        raise ValueError(f'Failed to run SQL query, error: {result.message}')

File: src/tools/sql.ts
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';

import { createKeboolaClients } from '@/clients/keboola';
import { createRawClient } from '@/clients/raw';
import { deriveServiceUrls } from '@/clients/urls';
import type { Config } from '@/config';
import { logger } from '@/logger';
import { registerTool } from '@/mcp/tool';
import { type SqlSelectData, WorkspaceManager } from '@/workspace';

// Ported from tools/sql.py.

const MAX_ROWS = 1_000;
const MAX_CHARS = 50_000;

const QUERY_DATA_DESCRIPTION = `
    Executes an SQL SELECT query to get the data from the underlying database.

    BEFORE QUERYING:
    * Always verify the table has a non-null fullyQualifiedName from get_tables tool.
      If it does not, the table is not SQL-accessible from this workspace — do not attempt the query and inform user.

    CRITICAL SQL REQUIREMENTS:

    * ALWAYS check the SQL dialect before constructing queries.
    * Do not include any comments in the SQL code
    * Use delimited identifiers and FQN format for the current SQL dialect.

    TABLE AND COLUMN REFERENCES:
    * Always use fully qualified table names in the exact FQN format provided by table information tools
    * Follow the identifier structure exactly as shown by table info tools for the current SQL dialect
    * Always use delimited identifiers when referring to table columns

    CTE (WITH CLAUSE) RULES:
    * ALL column references in main query MUST match exact case used in the CTE
    * If you alias a column in a CTE, reference it under the aliased name in the subsequent queries
    * Define all column aliases explicitly in CTEs
    * Use delimited identifiers in both CTE definition and references to preserve case

    FUNCTION COMPATIBILITY:
    * Check data types before using date functions (DATE_TRUNC, EXTRACT require proper date/timestamp types)
    * Cast VARCHAR columns to appropriate types before using in date/numeric functions

    ERROR PREVENTION:
    * Never pass empty strings ('') where numeric or date values are expected
    * Use NULLIF or CASE statements to handle empty values
    * Always use TRY_CAST or similar safe casting functions when converting data types
    * Check for division by zero using NULLIF(denominator, 0)
    * Always use the LIMIT clause in your SELECT statements when fetching data. There are hard limits imposed
      by this tool on the maximum number of rows that can be fetched and the maximum number of characters.
      The tool will truncate the data if those limits are exceeded.

    DATA VALIDATION:
    * When querying columns with categorical values, use query_data tool to inspect distinct values beforehand
    * Ensure valid filtering by checking actual data values first
    `;

/**
 * Serializes rows to CSV, matching Python's `csv.DictWriter` defaults:
 * comma delimiter, `\r\n` line terminator, minimal quoting (quote a field only
 * when it contains the delimiter, a quote, CR, or LF; double embedded quotes).
 */
const toCsv = (data: SqlSelectData): string => {
  const needsQuote = (field: string): boolean =>
    field.includes(',') || field.includes('"') || field.includes('\n') || field.includes('\r');
  const formatField = (value: unknown): string => {
    const s = value === null || value === undefined ? '' : String(value);
    return needsQuote(s) ? `"${s.replaceAll('"', '""')}"` : s;
  };
  const lines: string[] = [];
  lines.push(data.columns.map(formatField).join(','));
  for (const row of data.rows) {
    lines.push(data.columns.map((col) => formatField(row[col])).join(','));
  }
  return lines.map((line) => `${line}\r\n`).join('');
};

/**
 * Builds the WorkspaceManager from the resolved config. The query service /
 * workspace-discovery clients are built locally (the shared `createKeboolaClients`
 * does not expose a query service client), rooted at `query.<suffix>`.
 */
const createWorkspaceManager = async (config: Config): Promise<WorkspaceManager> => {
  const clients = createKeboolaClients(config);
  const urls = deriveServiceUrls(config.storageApiUrl ?? '');
  // Query Service host: no dedicated key in deriveServiceUrls — derive `query.<suffix>`
  // from the storage host, matching the Python `query.<hostname_suffix>` derivation.
  const suffix = new URL(urls.storage).hostname.slice('connection.'.length);
  const queryServiceUrl = `https://query.${suffix}`;

  // Query Service prefers the OAuth bearer token, falling back to the SAPI token.
  const queryServiceToken = config.bearerToken
    ? `Bearer ${config.bearerToken}`
    : (config.storageToken ?? '');

  // A production-branch raw storage client for the legacy / default-branch fallback path.
  const storageToken = config.bearerToken
    ? `Bearer ${config.bearerToken}`
    : (config.storageToken ?? '');
  const makeProdRawStorage = () =>
    createRawClient({ baseUrl: `${urls.storage}/v2/storage`, token: storageToken });

  return WorkspaceManager.create(config, {
    rawStorage: clients.rawStorage,
    makeProdRawStorage,
    queryServiceUrl,
    queryServiceToken,
  });
};

export const registerSqlTools = (server: McpServer, config: Config): void => {
  registerTool(server, {
    name: 'query_data',
    title: 'Query data',
    description: QUERY_DATA_DESCRIPTION,
    annotations: { readOnlyHint: true },
    inputSchema: {
      sql_query: z.string().describe('SQL SELECT query to run.'),
      query_name: z
        .string()
        .describe(
          'A concise, human-readable name for this query based on its purpose and what data it retrieves. ' +
            'Use normal words with spaces (e.g., "Customer Orders Last Month", "Top Selling Products", ' +
            '"User Activity Summary").',
        ),
    },
    handler: async (args) => {
      const workspaceManager = await createWorkspaceManager(config);

      const result = await workspaceManager.executeQuery(args.sql_query, {
        maxRows: MAX_ROWS,
        maxChars: MAX_CHARS,
      });

      if (result.status === 'ok') {
        logger.info(
          [`Query "${args.query_name}" executed successfully.`, result.message]
            .filter(Boolean)
            .join(' '),
        );
        const data: SqlSelectData = result.data
          ? result.data
          : // Non-SELECT query (should not happen for this SELECT-only tool).
            { columns: ['message'], rows: [{ message: result.message }] };

        return {
          query_name: args.query_name,
          csv_data: toCsv(data),
          message: result.message ?? null,
        };
      }

      // Surface cancellation cleanly without the generic "Failed to run SQL query" prefix.
      if (result.message === 'Query was cancelled') {
        logger.info(`Query "${args.query_name}" was cancelled.`);
        throw new Error('Query was cancelled');
      }
      logger.warn([`Query "${args.query_name}" failed.`, result.message].filter(Boolean).join(' '));
      throw new Error(`Failed to run SQL query, error: ${result.message}`);
    },
  });
};
